src/transformers/cache_utils.py

```diff
@@ -1089,9 +1097,21 @@ def update(
         Return:
             A tuple containing the updated key and value states.
         """
+
+        for state_str, state_device, self_state_device in [
+            ("key_states", key_states.device, self.key_cache[layer_idx].device),
+            ("value_states", value_states.device, self.value_cache[layer_idx].device),
+        ]:
+            if state_device != self_state_device:
+                raise ValueError(
+                    f"Computed {state_str} from layer {layer_idx} is on device {state_device} "
+                    f"whereas stored {state_str} is on device {self_state_device}. "
+                    f"If you are manually initializing the cache, make sure to pass the argument `layer_device_map` if you are using multi-gpu. "
+                    " Otherwise, you can just pass `cache_implementation` in `model.generate()` to correctly initialize the cache."
+                )
```

Comment:
I am really unsure this is worth it for us to run this at every forward pass. I know we want to help our users but would need to make sur it does not cost us anything

Reply:
We can also wrap `update` with a try/except, rather than using an if/else